eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -554,89 +556,137 @@ instance Semigroup ScriptResult where
 instance Monoid ScriptResult where
   mempty = Passes mempty
 
-data PlutusDebug
-  = PlutusDebugV1
-      CostModel
-      ExUnits
-      SBS.ShortByteString
-      [PV1.Data]
-      ProtVer
-  | PlutusDebugV2
-      CostModel
-      ExUnits
-      SBS.ShortByteString
-      [PV2.Data]
-      ProtVer
-  deriving (Eq, Generic, NoThunks)
-
--- | There is dummy Show instance for PlutusDebug intentionally, because it is too
+newtype PlutusData (l :: Language) = PlutusData {unPlutusData :: [PV1.Data]}
+  deriving (Eq)
+
+instance
+  forall (l :: Language).
+  (Typeable l) =>
+  ToCBOR (PlutusData l)
+  where
+  toCBOR (PlutusData d) = toCBOR d
+
+instance
+  forall (l :: Language).
+  (Typeable l) =>
+  FromCBOR (PlutusData l)
+  where
+  fromCBOR = PlutusData <$> fromCBOR
+
+data PlutusDebugLang (l :: Language) where
+  PlutusDebugLang ::
+    { pdSLanguage :: SLanguage l,
+      pdCostModel :: CostModel,
+      pdExUnits :: ExUnits,
+      pdSBS :: SBS.ShortByteString,
+      pdPlutusData :: PlutusData l,
+      pdProtVer :: ProtVer
+    } ->
+    PlutusDebugLang l
+
+-- | There is dummy Show instance for PlutusDebugLang intentionally, because it is too
 -- expensive and it will be too tempting to use it incorrectly. If needed for
 -- testing use 'StandaloneDeriving', otherwise define an efficient way to display
 -- this info.
-instance Show PlutusDebug where
+instance Show (PlutusDebugLang l) where
   show _ = "PlutusDebug Omitted"
 
-data PlutusError = PlutusErrorV1 PV1.EvaluationError | PlutusErrorV2 PV2.EvaluationError
+deriving instance forall (l :: Language). (Eq (SLanguage l)) => Eq (PlutusDebugLang l)
+
+deriving instance Generic (PlutusDebugLang l)
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l, ToCBOR (SLanguage l)) =>
+  ToCBOR (PlutusDebugLang l)
+  where
+  toCBOR (PlutusDebugLang slang costModel exUnits sbs pData protVer) =
+    encode $ Rec PlutusDebugLang !> To slang !> E encodeCostModel costModel !> To exUnits !> To sbs !> To pData !> To protVer
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  FromCBOR (PlutusDebugLang l)
+  where
+  fromCBOR = do
+    slang <- fromCBOR @(SLanguage l)
+    costModel <- decodeCostModel (fromSLanguage slang)
+    exUnits <- fromCBOR
+    sbs <- fromCBOR
+    pData <- fromCBOR
+    protVer <- fromCBOR
+    pure $ PlutusDebugLang slang costModel exUnits sbs pData protVer
```

Comment:
Unfortunately, this has to be the instance in order for it to be backwards compatible
```suggestion
  fromCBOR = decodeRecordSum "PlutusDebug" $ \t -> do
    let slang = isLanguage @(SLanguage l)
        lang = fromSLanguage slang
    when (fromEnum lang /= fromIntegral tag) $ 
      fail $ "Unexpected language: " ++ show  t
    slang <- fromCBOR 
    costModel <- decodeCostModel lang
    exUnits <- fromCBOR
    sbs <- fromCBOR
    pData <- fromCBOR
    protVer <- fromCBOR
    pure $ (1, PlutusDebugLang slang costModel exUnits sbs pData protVer)
```